control-services/cli/semaphore.py

```diff
@@ -61,8 +61,74 @@ def login(self, password=None):
 
         return True
 
-    def runTask(self):
-        return
+    def runTask(self, projectName, templateAlias, taskParams=None):
```

Comment:
I haven't looked into an active debug, but when I run this code, it's unclear to me how the arguments are supposed to be passed.

```console
control@control:~/vater$ vater task run -h
usage: vater.py task run [-h] projectName templateAlias [taskParams [taskParams ...]]

Runs a semaphore task in the specified project using the specified template

positional arguments:
  projectName    The name of the project where the task will execute (e.g. "Class 21012" or "Management")
  templateAlias  The alias of the task template to execute (e.g. "Create Range"). The template must be defined in the specified project
  taskParams     Parameters to be passed to the task's environment in key=value format (e.g. class=21012). Note that environment variables provided directly to the task template in the Semaphore UI
                 will take precedence over variables specified here. If experiencing inconsistencies with values provided here and the results in Semaphore, verify that the template does not have
                 hard-coded arguments that overwrite parameters you are providing at the command line

optional arguments:
  -h, --help     show this help message and exit
control@control:~/vater$ vater task run "Management" "Create Class" "class=21030 classSize=2"
Semaphore Password:
Task Created!
Task ID: 43
control@control:~/vater$ vater task run "Management" "Create Class" "class=21030 classSize=2"
Semaphore Password:
Task Created!
Task ID: 44
control@control:~/vater$ vater task run "Management" "Create Class" class=21030 classSize=2
Semaphore Password:
Task Created!
Task ID: 45
```

Task  44 failed with:
```
1:49:02 PM
Task 44 added to queue
1:49:07 PM
Preparing: 44
1:49:07 PM
Prepare task with template: Create Class
1:49:07 PM
Updating repository http://192.168.100.1:3000/333TRS/rous
1:49:07 PM
hint: Pulling without specifying how to reconcile divergent branches is
1:49:07 PM
hint: discouraged. You can squelch this message by running one of the following
1:49:07 PM
hint: commands sometime before your next pull:
1:49:07 PM
hint:
1:49:07 PM
hint: git config pull.rebase false # merge (the default strategy)
1:49:07 PM
hint: git config pull.rebase true # rebase
1:49:07 PM
hint: git config pull.ff only # fast-forward only
1:49:07 PM
hint:
1:49:07 PM
hint: You can replace "git config" with "git config --global" to set a default
1:49:07 PM
hint: preference for all repositories. You can also pass --rebase, --no-rebase,
1:49:07 PM
hint: or --ff-only on the command line to override the configured default per
1:49:07 PM
hint: invocation.
1:49:07 PM
From http://192.168.100.1:3000/333TRS/rous
1:49:07 PM
* branch master -> FETCH_HEAD
1:49:07 PM
Already up to date.
1:49:07 PM
Get current commit hash
1:49:07 PM
Get current commit message
1:49:07 PM
installing static inventory
1:49:07 PM
No roles/requirements.yml file found. Skip galaxy install process.
1:49:12 PM
Started: 44
1:49:12 PM
Run task with template: Create Class
1:49:13 PM
1:49:13 PM
PLAY [localhost] ***************************************************************
1:49:13 PM
1:49:13 PM
TASK [Gathering Facts] *********************************************************
1:49:15 PM
ok: [localhost]
1:49:15 PM
1:49:15 PM
TASK [Display token] ***********************************************************
1:49:15 PM
ok: [localhost] =>
1:49:15 PM
msg: Using token dxv1asb4zq_wwxple2vua2uiydzlvmbjxbydm0j9xrw=
1:49:15 PM
1:49:15 PM
TASK [Check Semaphore API is available] ****************************************
1:49:16 PM
ok: [localhost]
1:49:16 PM
1:49:16 PM
TASK [Create new project for Class 21030 classSize] ****************************
1:49:16 PM
ok: [localhost]
1:49:16 PM
1:49:16 PM
TASK [debug] *******************************************************************
1:49:16 PM
ok: [localhost] =>
1:49:16 PM
project:
1:49:16 PM
changed: false
1:49:16 PM
connection: close
1:49:16 PM
content: |-
1:49:16 PM
{"id":10,"name":"Class 21030 classSize","created":"2022-01-11T19:49:16.34936472Z","alert":false,"alert_chat":""}
1:49:16 PM
content_length: '113'
1:49:16 PM
content_type: application/json
1:49:16 PM
cookies: {}
1:49:16 PM
cookies_string: ''
1:49:16 PM
date: Tue, 11 Jan 2022 19:49:16 GMT
1:49:16 PM
elapsed: 0
1:49:16 PM
failed: false
1:49:16 PM
json:
1:49:16 PM
alert: false
1:49:16 PM
alert_chat: ''
1:49:16 PM
created: '2022-01-11T19:49:16.34936472Z'
1:49:16 PM
id: 10
1:49:16 PM
name: Class 21030 classSize
1:49:16 PM
msg: OK (113 bytes)
1:49:16 PM
redirected: false
1:49:16 PM
status: 201
1:49:16 PM
url: http://192.168.100.1:4000/api/projects
1:49:16 PM
1:49:16 PM
TASK [Setup New Class Project] *************************************************
1:49:17 PM
fatal: [localhost]: FAILED! =>
1:49:17 PM
msg: |-
1:49:17 PM
The task includes an option with an undefined variable. The error was: 'classSize' is undefined
1:49:17 PM
1:49:17 PM
The error appears to be in '/tmp/semaphore/repository_1_1/tasks/createClass.yml': line 34, column 5, but may
1:49:17 PM
be elsewhere in the file depending on the exact syntax problem.
1:49:17 PM
1:49:17 PM
The offending line appears to be:
1:49:17 PM
1:49:17 PM
1:49:17 PM
- name: Setup New Class Project
1:49:17 PM
^ here
1:49:17 PM
1:49:17 PM
PLAY RECAP *********************************************************************
1:49:17 PM
localhost : ok=5 changed=0 unreachable=0 failed=1 skipped=0 rescued=0 ignored=0
1:49:17 PM
1:49:17 PM
Running playbook failed: exit status 2
```

Task 43 created the same project again with behavior described where GUI command line arguments overwrote the ones passed via CLI.

Task 45 created:

![image](https://user-images.githubusercontent.com/55300322/149012083-2c890cdf-08c7-4b54-a7b6-252586a1b68a.png)

With nothing configured in the project.

Log from semaphore:
```
1:50:36 PM
Task 45 added to queue
1:50:37 PM
Preparing: 45
1:50:37 PM
Prepare task with template: Create Class
1:50:37 PM
Updating repository http://192.168.100.1:3000/333TRS/rous
1:50:37 PM
hint: Pulling without specifying how to reconcile divergent branches is
1:50:37 PM
hint: discouraged. You can squelch this message by running one of the following
1:50:37 PM
hint: commands sometime before your next pull:
1:50:37 PM
hint:
1:50:37 PM
hint: git config pull.rebase false # merge (the default strategy)
1:50:37 PM
hint: git config pull.rebase true # rebase
1:50:37 PM
hint: git config pull.ff only # fast-forward only
1:50:37 PM
hint:
1:50:37 PM
hint: You can replace "git config" with "git config --global" to set a default
1:50:37 PM
hint: preference for all repositories. You can also pass --rebase, --no-rebase,
1:50:37 PM
hint: or --ff-only on the command line to override the configured default per
1:50:37 PM
hint: invocation.
1:50:37 PM
From http://192.168.100.1:3000/333TRS/rous
1:50:37 PM
* branch master -> FETCH_HEAD
1:50:37 PM
Already up to date.
1:50:37 PM
Get current commit hash
1:50:37 PM
Get current commit message
1:50:37 PM
installing static inventory
1:50:37 PM
No roles/requirements.yml file found. Skip galaxy install process.
1:50:42 PM
Started: 45
1:50:42 PM
Run task with template: Create Class
1:50:43 PM
1:50:43 PM
PLAY [localhost] ***************************************************************
1:50:43 PM
1:50:43 PM
TASK [Gathering Facts] *********************************************************
1:50:45 PM
ok: [localhost]
1:50:45 PM
1:50:45 PM
TASK [Display token] ***********************************************************
1:50:45 PM
ok: [localhost] =>
1:50:45 PM
msg: Using token dxv1asb4zq_wwxple2vua2uiydzlvmbjxbydm0j9xrw=
1:50:45 PM
1:50:45 PM
TASK [Check Semaphore API is available] ****************************************
1:50:46 PM
ok: [localhost]
1:50:46 PM
1:50:46 PM
TASK [Create new project for Class 21030] **************************************
1:50:46 PM
ok: [localhost]
1:50:46 PM
1:50:46 PM
TASK [debug] *******************************************************************
1:50:46 PM
ok: [localhost] =>
1:50:46 PM
project:
1:50:46 PM
changed: false
1:50:46 PM
connection: close
1:50:46 PM
content: |-
1:50:46 PM
{"id":11,"name":"Class 21030","created":"2022-01-11T19:50:46.332051671Z","alert":false,"alert_chat":""}
1:50:46 PM
content_length: '104'
1:50:46 PM
content_type: application/json
1:50:46 PM
cookies: {}
1:50:46 PM
cookies_string: ''
1:50:46 PM
date: Tue, 11 Jan 2022 19:50:46 GMT
1:50:46 PM
elapsed: 0
1:50:46 PM
failed: false
1:50:46 PM
json:
1:50:46 PM
alert: false
1:50:46 PM
alert_chat: ''
1:50:46 PM
created: '2022-01-11T19:50:46.332051671Z'
1:50:46 PM
id: 11
1:50:46 PM
name: Class 21030
1:50:46 PM
msg: OK (104 bytes)
1:50:46 PM
redirected: false
1:50:46 PM
status: 201
1:50:46 PM
url: http://192.168.100.1:4000/api/projects
1:50:46 PM
1:50:46 PM
TASK [Setup New Class Project] *************************************************
1:50:48 PM
changed: [localhost]
1:50:48 PM
1:50:48 PM
TASK [debug] *******************************************************************
1:50:48 PM
ok: [localhost] =>
1:50:48 PM
setupNewClass.stdout: |-
1:50:48 PM
[{'id': 28,
1:50:48 PM
'login_password': {'login': '', 'password': ''},
1:50:48 PM
'name': 'NoneKey',
1:50:48 PM
'override_secret': False,
1:50:48 PM
'project_id': 11,
1:50:48 PM
'removed': False,
1:50:48 PM
'ssh': {'login': '', 'passphrase': '', 'private_key': ''},
1:50:48 PM
'type': 'None'}]
1:50:48 PM
Found ID: 28
1:50:48 PM
[{'id': 29,
1:50:48 PM
'login_password': {'login': '', 'password': ''},
1:50:48 PM
'name': 'control',
1:50:48 PM
'override_secret': False,
1:50:48 PM
'project_id': 11,
1:50:48 PM
'removed': False,
1:50:48 PM
'ssh': {'login': '', 'passphrase': '', 'private_key': ''},
1:50:48 PM
'type': 'ssh'},
1:50:48 PM
{'id': 28,
1:50:48 PM
'login_password': {'login': '', 'password': ''},
1:50:48 PM
'name': 'NoneKey',
1:50:48 PM
'override_secret': False,
1:50:48 PM
'project_id': 11,
1:50:48 PM
'removed': False,
1:50:48 PM
'ssh': {'login': '', 'passphrase': '', 'private_key': ''},
1:50:48 PM
'type': 'None'}]
1:50:48 PM
Found ID: 29
1:50:48 PM
[{'id': 29,
1:50:48 PM
'login_password': {'login': '', 'password': ''},
1:50:48 PM
'name': 'control',
1:50:48 PM
'override_secret': False,
1:50:48 PM
'project_id': 11,
1:50:48 PM
'removed': False,
1:50:48 PM
'ssh': {'login': '', 'passphrase': '', 'private_key': ''},
1:50:48 PM
'type': 'ssh'},
1:50:48 PM
{'id': 28,
1:50:48 PM
'login_password': {'login': '', 'password': ''},
1:50:48 PM
'name': 'NoneKey',
1:50:48 PM
'override_secret': False,
1:50:48 PM
'project_id': 11,
1:50:48 PM
'removed': False,
1:50:48 PM
'ssh': {'login': '', 'passphrase': '', 'private_key': ''},
1:50:48 PM
'type': 'None'},
1:50:48 PM
{'id': 30,
1:50:48 PM
'login_password': {'login': '', 'password': ''},
1:50:48 PM
'name': 'NoneKeyLP',
1:50:48 PM
'override_secret': False,
1:50:48 PM
'project_id': 11,
1:50:48 PM
'removed': False,
1:50:48 PM
'ssh': {'login': '', 'passphrase': '', 'private_key': ''},
1:50:48 PM
'type': 'login_password'}]
1:50:48 PM
Found ID: 30
1:50:48 PM
[{'git_url': 'http://192.168.100.1:3000/333TRS/rous',
1:50:48 PM
'id': 10,
1:50:48 PM
'name': 'Playbooks',
1:50:48 PM
'project_id': 11,
1:50:48 PM
'removed': False,
1:50:48 PM
'ssh_key_id': 28}]
1:50:48 PM
Found ID: 10
1:50:48 PM
[{'become_key_id': None,
1:50:48 PM
'id': 12,
1:50:48 PM
'inventory': '[control]\n192.168.100.1',
1:50:48 PM
'name': 'control',
1:50:48 PM
'project_id': 11,
1:50:48 PM
'removed': False,
1:50:48 PM
'ssh_key_id': 29,
1:50:48 PM
'type': 'static'}]
1:50:48 PM
Found ID: 12
1:50:48 PM
[{'id': 10,
1:50:48 PM
'json': '{"classNumber": "21030","classSize": 2,"ansiblePathInRepository": '
1:50:48 PM
'"tasks","vmPathInRepository": '
1:50:48 PM
'"/home/control/rous/vms/","terraformPathInRepositoryOnControl": '
1:50:48 PM
'"/home/control/rous/terraform/"}',
1:50:48 PM
'name': 'Env',
1:50:48 PM
'password': None,
1:50:48 PM
'project_id': 11,
1:50:48 PM
'removed': False}]
1:50:48 PM
Found ID: 10
1:50:48 PM
1:50:48 PM
PLAY RECAP *********************************************************************
1:50:48 PM
localhost : ok=7 changed=1 unreachable=0 failed=0 skipped=0 rescued=0 ignored=0
1:50:48 PM
```

Probably need:
- Clearer help message from the `*nargs`
- More aggressive error checking in `semaphore.py` for expected format so I can't enter information incorrectly

Reply:
Thinking about this more, I remember you said:

> This is a good idea and was what I originally planned to do, but unfortunately Semaphore does not give you a way to overwrite those arguments provided in the template

But can we query for the arguments in `parser.py` and then enter them as `choices` for the parser instead of `nargs`? This would clear up the variable names, ensure the variables passed were the variables expected, and allow us to rely on `argparse` for a lot of the error handling / input validation.

If we can't query for them we probably want some way to grab them. I believe ansible command line arguments specified with the `-e` in the GUI overwrite `environment` variables, so maybe the solution is to leave place holders in the `environment` with a special `key=value` where `value` is some kind of static value like `COMMANDLINE`. Then we can query for them and know what the expected values are so `parser.py` can be specific about what it's willing to expect per task template.

This method would also give us more flexibility in the future when defining different task templates.

Reply:
In response to:
> But can we query for the arguments in parser.py and then enter them as choices for the parser instead of nargs?

Yes we can, however I would push back on this and say this is a feature enhancement rather than a necessary aspect of the functionality. I'll agree that the help message should be improved so that, at a minimum, it is clear how to correctly pass key=value pairs to the command line. However, I think it is also important to avoid scope creep. I feel it would be more appropriate to create a separate issue to the effect of "implement available choices for command line tasks." 

Additionally, as I am currently looking into implementing the functionality we are looking for in Semaphore, I am loath to add _more_ to our code that makes us dependent on this "workaround" implementation (but this is just an afterthought - my main concern is scope creep)

Reply:
One thing to think about: the `help` statement _does_ make it clear, as is, that the task parameters are an array of separate key-value pairs:
`usage: vater.py task run [-h] projectName templateAlias **[taskParams [taskParams ...]]**`

In your example, you entered the taskParams as a single string, not an array:
`control@control:~/vater$ vater task run "Management" "Create Class" "class=21030 classSize=2"`

Perhaps a clearer example in the `help` statement might reduce the tendency for misinterpretation here.

Reply:
> I'm going to push back on this and say this is a feature enhancement rather than a necessary aspect of the functionality. 

1. You're right that `choices` might be scope creep if you think it's complex-- really not concerned with how it happens, but let's do some basic error handling for incorrectly passing arguments in the CLI. 

2. Task 45 looks like it has a passable call with the CLI, but the output photo in `Semaphore` seems wrong. Did I use the CLI wrong or can you reproduce?

![image](https://user-images.githubusercontent.com/55300322/149192584-9cd9b911-b4a2-4551-9aea-1876355621f8.png)

```
control@control:~/vater$ vater task run "Management" "Create Class" class=21030 classSize=2
Semaphore Password:
Task Created!
Task ID: 45
```

Looks similar minus the quotes around `"Management"`.

Reply:
What about Task 45 looks incorrect? I'm seeing that the project was created via the API with the expected values, per this line:
```console
{"id":11,"name":"Class 21030","created":"2022-01-11T19:50:46.332051671Z","alert":false,"alert_chat":""}
```
Environment is also being created correctly:
```console
'json': '{"classNumber": "21030","classSize": 2,"ansiblePathInRepository": '
```
That seems like pretty solid evidence the variables were taken in correctly...are you sure the problem you saw in the GUI corresponds to task 45?

Reply:
Good point, that makes sense, I'll look when I get home. I'll run again with different class number.

Reply:
Updated the help message to include an example command so it is more clear how the arguments are supposed to be passed:
![image](https://user-images.githubusercontent.com/24918678/150408984-73cffb8f-a47c-4cce-9d3f-0967586c764d.png)